### PR TITLE
introducing gamma-jet tree producer task

### PR DESCRIPTION
**Pull request by fjonasALICE**
- addition of new workflow which allows to store a tree containing photons and jets, including isolation information for the jet. Ideal derived data to perform any gamma-jet analysis

### CODE REVIEW

File: PWGJE/DataModel/GammaJetAnalysisTree.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+DECLARE_SOA_COLUMN(TMdeltaPhi, tmdeltaphi, float);                     //! delta phi between cluster and closest match
+DECLARE_SOA_COLUMN(TMdeltaEta, tmdeltaeta, float);                     //! delta eta between cluster and closest match
+DECLARE_SOA_COLUMN(TMtrackP, tmtrackp, float);                         //! track momentum of closest match, -1 if no match found
+} // namespace gjgamma
+DECLARE_SOA_TABLE(GjGammas, "AOD", "GJGAMMA",
+                  gjgamma::ColId, gjgamma::Energy, gjgamma::Eta, gjgamma::Phi, gjgamma::M02, gjgamma::M20, gjgamma::NCells, gjgamma::Time, gjgamma::IsExotic, gjgamma::DistanceToBadChannel, gjgamma::NLM, gjgamma::IsoRaw, gjgamma::PerpConeRho, gjgamma::TMdeltaPhi, gjgamma::TMdeltaEta, gjgamma::TMtrackP)
+namespace gjevent
+{ // TODO add rho
+// DECLARE_SOA_INDEX_COLUMN(JCollision, collision);
+DECLARE_SOA_COLUMN(ColId, colid, int);
```
Comment (nzardosh) [inline PWGJE/DataModel/GammaJetAnalysisTree.h:51]: Why do need this CollId?

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+  // photon candidates
+  Produces<aod::GjChargedJets> chargedJetsTable;
+  Produces<aod::GjEvents> eventsTable;
+  Produces<aod::GjGammas> gammasTable;
+
+  HistogramRegistry mHistograms{"GammaJetTreeProducerHisto"};
+
+  // configure zorro
+  Service<o2::ccdb::BasicCCDBManager> ccdb;
+  Zorro zorro;
```
Comment (nzardosh) [inline PWGJE/Tasks/gammajettreeproducer.cxx:67]: here you should not use zorro anymore and get the decisions directly from the triggerSel coloum in JCollisions. You can find examples in jetFinderQA

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+  // ---------------------
+  // Processing functions
+  // ---------------------
+
+  // define cluster filter. It selects only those clusters which are of the type
+  // sadly passing of the string at runtime is not possible for technical region so cluster definition is
+  // an integer instead
+  Filter clusterDefinitionSelection = (o2::aod::jcluster::definition == mClusterDefinition);
+  // Process clusters
+  void processClusters(soa::Join<JetCollisions, aod::BkgChargedRhos, aod::JCollisionBCs>::iterator const& collision, aod::JBCs const&, selectedClusters const& clusters, JetTracks const& tracks)
```
Comment (nzardosh) [inline PWGJE/Tasks/gammajettreeproducer.cxx:239]: you can remove JBCs once you remove zorro

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+  {
+    // get bc
+    auto bc = collision.bc_as<aod::JBCs>();
+    if (!isEventAccepted(collision, bc)) {
+      return;
+    }
+
+    // loop over charged jets
+    for (auto jet : chargedJets) {
+      if (jet.pt() < jetPtMin)
```
Comment (nzardosh) [inline PWGJE/Tasks/gammajettreeproducer.cxx:307]: you can also add this as a filter to the jet table if you want

File: PWGJE/DataModel/GammaJetAnalysisTree.h (hunk 1 above)
Comment (nzardosh) [reply PWGJE/DataModel/GammaJetAnalysisTree.h:51]: You could instead make it a coloumn that saves the index of the GjEvents table and then fill that (you would need to create a map between this table and jcollisions)

**Comment by fjonasALICE** [comment]
Hi @nzardosh I implemented almost all your comments, except for the filter for jet pt, which i will implement later (but i think it is not crucial)

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+      return false;
+    }
+
+    return true;
+  }
+
+  bool isEventAccepted(const auto& collision)
+  {
+
+    if (collision.posZ() > mVertexCut) {
```
Comment (nzardosh) [inline PWGJE/Tasks/gammajettreeproducer.cxx:142]: you should take the absolute value of collision.posZ()

File: PWGJE/Tasks/gammajettreeproducer.cxx (hunk 3 above)
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: removed JBCs

File: PWGJE/Tasks/CMakeLists.txt [hunk 6]
```diff
@@ -164,6 +164,11 @@ if(FastJet_FOUND)
                         SOURCES fulljetspectrapp.cxx
                         PUBLIC_LINK_LIBRARIES O2::Framework O2Physics::PWGJECore O2Physics::AnalysisCore
                         COMPONENT_NAME Analysis)
+    o2physics_add_dpl_workflow(gamma-jet-tree-producer
+                        SOURCES gammajettreeproducer.cxx
+                        PUBLIC_LINK_LIBRARIES O2::Framework O2::EMCALBase O2::EMCALCalib O2Physics::PWGJECore O2Physics::AnalysisCore
+                        O2Physics::EventFilteringUtils
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/CMakeLists.txt]: both removed

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+// \struct GammaJetTreeProducer
+/// \brief Task to produce a tree for gamma-jet analysis, including photons (and information of isolation) and charged and full jets
+/// \author Florian Jonas <florian.jonas@cern.ch>, UC Berkeley/LBNL
+/// \since 02.08.2024
+///
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+using bcEvSelIt = o2::soa::Join<o2::aod::BCs, o2::aod::BcSels>::iterator;
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: removed

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+  Configurable<std::string> triggerMasks{"triggerMasks", "", "possible JE Trigger masks: fJetChLowPt,fJetChHighPt,fTrackLowPt,fTrackHighPt,fJetD0ChLowPt,fJetD0ChHighPt,fJetLcChLowPt,fJetLcChHighPt,fEMCALReadout,fJetFullHighPt,fJetFullLowPt,fJetNeutralHighPt,fJetNeutralLowPt,fGammaVeryHighPtEMCAL,fGammaVeryHighPtDCAL,fGammaHighPtEMCAL,fGammaHighPtDCAL,fGammaLowPtEMCAL,fGammaLowPtDCAL,fGammaVeryLowPtEMCAL,fGammaVeryLowPtDCAL"};
+  Configurable<std::string>
+    trackSelections{"trackSelections", "globalTracks", "set track selections"};
+  Configurable<float> trackMinPt{"trackMinPt", 0.15, "minimum track pT cut"};
+  Configurable<float> jetPtMin{"jetPtMin", 5.0, "minimum jet pT cut"};
+  Configurable<float> jetR{"jetR", 0.4, "jet resolution parameter"};
+  Configurable<float> isoR{"isoR", 0.4, "isolation cone radius"};
+
+  // cluster cuts
+  Configurable<int> mClusterDefinition{"clusterDefinition", 10, "cluster definition to be selected, e.g. 10=kV3Default"};
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx:79]: prefer to keep consistent with other tasks for now, will change eventually

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+  // Helper functions
+  // ---------------------
+  template <typename T>
+  double deltaR(T const& eta1, T const& phi1, T const& eta2, T const& phi2)
+  {
+    // make sure both phi1 and phi2 are 0 to 2pi
+
+    double dPhi = TVector2::Phi_0_2pi(phi1) - TVector2::Phi_0_2pi(phi2);
+    if (abs(dPhi) > M_PI) {
+      dPhi = 2 * M_PI - abs(dPhi);
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: replaced with deltaR function. I did some additional fixes to deltaR function in jetutilities that should improve convenience and safety

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+    return true;
+  }
+
+  bool isEventAccepted(const auto& collision)
+  {
+
+    if (collision.posZ() > mVertexCut) {
+      return false;
+    }
+    if (!jetderiveddatautilities::selectCollision(collision, eventSelection) || !jetderiveddatautilities::selectTrigger(collision, triggerMaskBits)) {
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: will leave for now but will clarify with Nicolas Strangmann for a future update

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+      //   }
+      // }
+
+      // // for compression reasons make dPhi and dEta 0 if no match is found
+      // if (p == -1) {
+      //   dPhi = 0;
+      //   dEta = 0;
+      // }
+
+      gammasTable(collisionMapping[collision.globalIndex()], cluster.energy(), cluster.eta(), cluster.phi(), cluster.m02(), cluster.m20(), cluster.nCells(), cluster.time(), cluster.isExotic(), cluster.distanceToBadChannel(), cluster.nlm(), isoraw, perpconerho, dPhi, dEta, p);
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: fixed and changed to unordered map to improve performance

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    // dummy loop over tracks
+    for (auto track : tracks) {
+      mHistograms.fill(HIST("trackPt"), track.pt());
+    }
+  }
+  PROCESS_SWITCH(GammaJetTreeProducer, processClusters, "Process EMCal clusters", true);
+
+  Filter jetCuts = aod::jet::pt > jetPtMin&& aod::jet::r == nround(jetR.node() * 100.0f);
+  // Process charged jets
+  void processChargedJets(soa::Join<JetCollisions, aod::BkgChargedRhos, aod::JCollisionBCs>::iterator const& collision, aod::JBCs const&, soa::Filtered<soa::Join<aod::ChargedJets, aod::ChargedJetConstituents>> const& chargedJets, JetTracks const&)
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: removed

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+      if (jet.pt() < jetPtMin)
+        continue;
+      ushort nconst = 0;
+      // loop over constituents
+      for (auto& constituent : jet.template tracks_as<JetTracks>()) {
+        mHistograms.fill(HIST("chjetpt_vs_constpt"), jet.pt(), constituent.pt());
+        nconst++;
+      }
+
+      chargedJetsTable(collisionMapping[collision.globalIndex()], jet.pt(), jet.eta(), jet.phi(), jet.energy(), jet.mass(), jet.area(), nconst);
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: fixed

File: PWGJE/Tasks/gammajettreeproducer.cxx [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+      // fill histograms
+      mHistograms.fill(HIST("chjetPt"), jet.pt());
+    }
+  }
+  PROCESS_SWITCH(GammaJetTreeProducer, processChargedJets, "Process charged jets", true);
+};
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{
+    adaptAnalysisTask<GammaJetTreeProducer>(cfgc, TaskName{"gamma-jet-tree-producer"}, SetDefaultProcesses{{{"processClearMaps", true}, {"processClusters", true}, {"processChargedJets", true}}})};
```
[the remark this answers is not in the document]
Comment (fjonasALICE) [reply PWGJE/Tasks/gammajettreeproducer.cxx]: fixed

**Review by nzardosh — APPROVED** [review]

**Pull request merged**